src/erasure-code/clay/ErasureCodeClay.cc

```diff
@@ -0,0 +1,1007 @@
+    return ErasureCode::minimum_to_decode(want_to_read, available, minimum);
+  }
+}
+
+int ErasureCodeClay::decode(const set<int> &want_to_read,
+                const map<int, bufferlist> &chunks,
+                map<int, bufferlist> *decoded, int chunk_size){
+  set<int> avail;
+  for (map<int, bufferlist>::const_iterator i = chunks.begin();
+      i != chunks.end(); ++i) {
+    avail.insert(i->first);
+  }
+
+  if (is_repair(want_to_read, avail)) {
+    return repair(want_to_read, chunks, decoded);
+  } else {
+    return ErasureCode::_decode(want_to_read, chunks, decoded);
+  }
+}
+
+void p(const set<int> &s) { cerr << s; } // for gdb
+
+int ErasureCodeClay::encode_chunks(const set<int> &want_to_encode,
+                                       map<int, bufferlist> *encoded)
+{
+  map<int, bufferlist> chunks;
+  set<int> parity_chunks;
+  int chunk_size = (*encoded)[0].length();
+  
+  for (int i = 0; i < k + m; i++) {
+    if (i < k) {
+      chunks[i] = (*encoded)[i];
+    } else {
+      chunks[i+nu] = (*encoded)[i];
+	  parity_chunks.insert(i+nu);
+    }
+  }
+  
+  for (int i = k; i < k+nu; i++) {
+	bufferptr buf(buffer::create_aligned(chunk_size, SIMD_ALIGN));
+    buf.zero();
+    chunks[i].push_back(std::move(buf));  
+  }
+  
+  int res = decode_layered(parity_chunks, &chunks);
+  for (int i = k ; i < k+nu; i++) {
+	//need to clean some of the intermediate chunks here!!
+	chunks[i].clear();
+  }
+  return res;
+}
+
+int ErasureCodeClay::decode_chunks(const set<int> &want_to_read,
+				  const map<int, bufferlist> &chunks,
+				  map<int, bufferlist> *decoded)
+{
+  set<int> erasures;
+  map<int, bufferlist> coded_chunks;
+  
+  for (int i = 0; i < k + m; i++) {
+	if (want_to_read.find(i) != want_to_read.end()) {
+      erasures.insert(i < k ? i : i+nu);
+	}
+	if (chunks.find(i) == chunks.end()) {
+	  assert((*decoded).find(i) != (*decoded).end());
+	  coded_chunks[i < k ? i : i+nu] = (*decoded)[i];
+	} else {
+      coded_chunks[i < k ? i : i+nu] = (chunks.find(i))->second;
+	}
+  }
+  int chunk_size = coded_chunks[0].length();
+  
+  for (int i = k; i < k+nu; i++) {
+	bufferptr buf(buffer::create_aligned(chunk_size, SIMD_ALIGN));
+    buf.zero();
+    coded_chunks[i].push_back(std::move(buf));  
+  }
+	
+  int res = decode_layered(erasures, &coded_chunks);
+  for (int i = k; i < k+nu; i++) {
+    coded_chunks[i].clear();  
+  }
+	
+  return res;
+}
+
+unsigned int ErasureCodeClay::get_alignment() const
+{
+	unsigned alignment = k*sub_chunk_no*w*sizeof(int);
+    if ((w*sizeof(int))%LARGEST_VECTOR_WORDSIZE)
+      alignment = k*sub_chunk_no*w*LARGEST_VECTOR_WORDSIZE;
+    return alignment;
+}
+
+int ErasureCodeClay::parse(ErasureCodeProfile &profile,
+			  ostream *ss)
+{
+ int err = 0;
+  err = ErasureCode::parse(profile, ss);
+  err |= to_int("k", profile, &k, DEFAULT_K, ss);
+  err |= to_int("m", profile, &m, DEFAULT_M, ss);
+  err |= to_int("m", profile, &m, DEFAULT_W, ss);
```

Comment:
Is there a mistake? I think that it should be to_int("w", profile, &w, DEFAULT_W, ss), not to_int("m", profile, &m, DEFAULT_W, ss).

Reply:
yes this is a typo.. it will be fixed. Thanks!